specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json

```diff
@@ -21156,6 +22197,228 @@
+          "description": "Specifies the content safety options. If omitted, the default content safety settings will be configured",
+          "x-nullable": true
+        },
+        "endpointState": {
+          "$ref": "#/definitions/ServerlessEndpointState",
+          "description": "The current state of the ServerlessEndpoint.",
+          "readOnly": true,
+          "x-ms-mutability": [
+            "read"
+          ]
+        },
+        "inferenceEndpoint": {
+          "$ref": "#/definitions/ServerlessInferenceEndpoint",
+          "description": "The inference uri to target when making requests against the serverless endpoint",
+          "readOnly": true,
+          "x-ms-mutability": [
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "marketplaceSubscriptionId": {
+          "description": "The MarketplaceSubscription ARM ID associated to this ServerlessEndpoint.",
+          "readOnly": true,
+          "type": "string",
+          "x-ms-mutability": [
+            "read"
+          ],
+          "x-nullable": true
+        },
+        "modelSettings": {
+          "$ref": "#/definitions/ModelSettings",
+          "description": "The model settings (model id) for the model being serviced on the ServerlessEndpoint.",
+          "x-nullable": true
+        },
+        "provisioningState": {
+          "$ref": "#/definitions/EndpointProvisioningState",
+          "description": "Provisioning state for the endpoint.",
+          "readOnly": true,
+          "x-ms-mutability": [
+            "read"
+          ]
+        }
+      },
+      "required": [
+        "authMode"
+      ],
+      "type": "object",
+      "x-ms-client-name": "ServerlessEndpointProperties"
+    },
+    "ServerlessEndpointState": {
+      "additionalProperties": false,
+      "description": "State of the Serverless Endpoint.",
+      "enum": [
+        "Unknown",
+        "Creating",
+        "Deleting",
+        "Suspending",
+        "Reinstating",
+        "Online",
+        "Suspended",
+        "CreationFailed",
+        "DeletionFailed"
+      ],
+      "type": "string",
+      "x-ms-enum": {
+        "modelAsString": true,
+        "name": "ServerlessEndpointState",
+        "values": [
+          {
+            "value": "Unknown"
+          },
+          {
+            "value": "Creating"
+          },
+          {
+            "value": "Deleting"
+          },
+          {
+            "value": "Suspending"
+          },
+          {
+            "value": "Reinstating"
+          },
+          {
+            "value": "Online"
+          },
+          {
+            "value": "Suspended"
+          },
+          {
+            "value": "CreationFailed"
+          },
+          {
+            "value": "DeletionFailed"
+          }
+        ]
+      }
+    },
+    "ServerlessEndpointStatus": {
+      "additionalProperties": false,
+      "properties": {
+        "metrics": {
```

Comment:
>metrics [](http://example.com/codeflow?start=9&length=7)

additionalProperties pattern is an ARM anti-pattern. It's best if you can schematize the properties. In some cases, if they are too variable or determined by an external source, you can consider using an array instead. Array is a much more ARM friendly pattern.

Reply:
The list of metrics will be model dependent, some of which may come from 3P ISVs etc. Hence it is modeled as a generic string dictionary.

The key thing here is that our service is not in control of all possible keys in this dictionary, because this data comes from customers (specifically, model providers).